File: electron/backend/modules_old/variants/variant.service.ts
/**
 * Variant Service
 * Contains business logic for variant operations
 */
import * as variantRepository from './variant.repository.js';
import * as inventoryRepository from '../inventory/inventory.repository.js';
import * as auditService from '../audit/audit.service.js';
import { transaction } from '../../shared/db/index.js';
import { buildImageUrl } from '../../shared/utils/index.js';
import { VariantQueryOptions } from './variant.repository.js';
import { Variant, INVENTORY_REASONS } from '../../../../types/index.js';
import { ValidationError } from '../../shared/errors/index.js';
import { logger } from '../../shared/utils/logger.js';

interface AddVariantInput {
    productId: number;
    name: string;
    sku?: string | null;
    price: number;
    cost_price: number;
    stock_alert_cap?: number;
    is_default?: boolean;
    status?: 'active' | 'inactive' | 'discontinued';
    stock?: number;
    userId: number;
}

interface UpdateVariantInput {
    id: number;
    name: string;
    sku?: string | null;
    price: number;
    cost_price: number;
    stock_alert_cap?: number;
    is_default?: boolean;
    status?: 'active' | 'inactive' | 'discontinued';
    stock?: number;
    userId: number;
}

/**
 * Add a variant to a product
 */
export function addVariant(input: AddVariantInput) {
    if (!input.productId || !input.name || input.price == null || input.cost_price == null || !input.userId) {
        throw new ValidationError('Product ID, name, price, cost_price, and userId are required');
    }

    if (input.price < 0 || input.cost_price < 0) {
        throw new ValidationError('Price and cost_price must be non-negative');
    }

    const tx = transaction(() => {
        const result = variantRepository.insertVariant(input.productId, {
            name: input.name,
            sku: input.sku,
            price: input.price,
            cost_price: input.cost_price,
            stock_alert_cap: input.stock_alert_cap ?? 10,
            is_default: input.is_default ? 1 : 0,
            status: input.status ?? 'active'
        });
        const variantId = Number(result.lastInsertRowid);

        auditService.logCreate(input.userId, 'variants', variantId, {
            product_id: input.productId,
            name: input.name,
            sku: input.sku,
            price: input.price,
            cost_price: input.cost_price
        });

        if (input.stock && input.stock > 0) {
            const lotResult = inventoryRepository.insertInventoryLot({
                variant_id: variantId,
                quantity: parseInt(String(input.stock), 10),
                unit_cost: input.cost_price
            });
            const lotId = Number(lotResult.lastInsertRowid);

            inventoryRepository.insertInventoryAdjustment({
                variant_id: variantId,
                lot_id: lotId,
                quantity_change: parseInt(String(input.stock), 10),
                reason: INVENTORY_REASONS.CONFIRM_RECEIVE,
                adjusted_by: input.userId
            });

            logger.info(`Initial stock ${input.stock} added for variant ${variantId}`);
        }

        return result;
    });
    return tx();
}

/**
 * Update a variant
 */
export function updateVariant(input: UpdateVariantInput) {
    if (!input.id || !input.name || input.price == null || input.cost_price == null || !input.userId) {
        throw new ValidationError('ID, name, price, cost_price, and userId are required');
    }

    if (input.price < 0 || input.cost_price < 0) {
        throw new ValidationError('Price and cost_price must be non-negative');
    }

    const existing = variantRepository.findVariantByIdSync(input.id);
    if (!existing) {
        throw new ValidationError('Variant not found');
    }

    const tx = transaction(() => {
        const result = variantRepository.updateVariantById({
            id: input.id,
            name: input.name,
            sku: input.sku,
            price: input.price,
            cost_price: input.cost_price,
            stock_alert_cap: input.stock_alert_cap ?? 10,
            is_default: input.is_default ? 1 : 0,
            status: input.status ?? 'active'
        });

        auditService.logUpdate(input.userId, 'variants', input.id, existing, {
            name: input.name,
            sku: input.sku,
            price: input.price,
            cost_price: input.cost_price
        });

        if (input.stock !== undefined) {
            const targetStock = parseInt(String(input.stock), 10);
            if (!isNaN(targetStock) && targetStock >= 0) {
                const currentStock = inventoryRepository.getStockByVariantId(input.id);
                const diff = targetStock - currentStock;

                if (diff !== 0) {
                    inventoryRepository.insertInventoryAdjustment({
                        variant_id: input.id,
                        quantity_change: diff,
                        reason: INVENTORY_REASONS.CORRECTION,
                        adjusted_by: input.userId
                    });
                    logger.info(`Stock adjusted for variant ${input.id}: ${diff > 0 ? '+' : ''}${diff}`);
                }
            }
        }
        return result;
    });
    return tx();
}

/**
 * Delete a variant (soft delete)
 */
export function deleteVariant(id: number, userId: number) {
    if (!id || !userId) {
        throw new ValidationError('ID and userId are required');
    }

    const existing = variantRepository.findVariantByIdSync(id);
    if (!existing) {
        throw new ValidationError('Variant not found');
    }

    const result = variantRepository.softDeleteVariant(id);
    
    auditService.logDelete(userId, 'variants', id, existing);
    logger.info(`Variant ${id} soft deleted by user ${userId}`);
    
    return result;
}

/**
 * Get variants with filtering, pagination and sorting
 */
export async function getVariants(params: VariantQueryOptions) {
    const result = await variantRepository.findVariants(params);

    return {
        ...result,
        variants: result.variants.map((variant: any) => ({
            ...variant,
            taxes: variant.taxes ? JSON.parse(variant.taxes) : [],
            imageUrl: buildImageUrl(variant.image_path)
        }))
    };
}

/**
 * Get variant statistics
 */
export async function getVariantStats() {
    return await variantRepository.getVariantStats();
}

File: src/main/java/com/possum/application/variants/VariantService.java
package com.possum.application.variants;

import com.possum.domain.exceptions.NotFoundException;
import com.possum.domain.exceptions.ValidationException;
import com.possum.domain.model.InventoryAdjustment;
import com.possum.domain.model.InventoryLot;
import com.possum.domain.model.Variant;
import com.possum.infrastructure.logging.LoggingConfig;
import com.possum.persistence.db.TransactionManager;
import com.possum.persistence.repositories.interfaces.AuditRepository;
import com.possum.persistence.repositories.interfaces.InventoryRepository;
import com.possum.persistence.repositories.interfaces.VariantRepository;
import com.possum.shared.dto.PagedResult;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

public class VariantService {
    private final VariantRepository variantRepository;
    private final InventoryRepository inventoryRepository;
    private final AuditRepository auditRepository;
    private final TransactionManager transactionManager;

    public VariantService(VariantRepository variantRepository,
                          InventoryRepository inventoryRepository,
                          AuditRepository auditRepository,
                          TransactionManager transactionManager) {
        this.variantRepository = variantRepository;
        this.inventoryRepository = inventoryRepository;
        this.auditRepository = auditRepository;
        this.transactionManager = transactionManager;
    }

    public long addVariant(AddVariantCommand command) {
        if (command.productId() == null || command.name() == null || command.price() == null || command.costPrice() == null || command.userId() == null) {
            throw new ValidationException("Product ID, name, price, cost_price, and userId are required");
        }

        if (command.price().compareTo(BigDecimal.ZERO) < 0 || command.costPrice().compareTo(BigDecimal.ZERO) < 0) {
            throw new ValidationException("Price and cost_price must be non-negative");
        }

        return transactionManager.runInTransaction(() -> addVariantWithoutTransaction(command));
    }

    public long addVariantWithoutTransaction(AddVariantCommand command) {
        Variant variant = new Variant(
                null,
                command.productId(),
                null,
                command.name(),
                command.sku(),
                command.price(),
                command.costPrice(),
                command.stockAlertCap() != null ? command.stockAlertCap() : 10,
                command.isDefault() != null && command.isDefault(),
                command.status() != null ? command.status() : "active",
                null, 
                null,
                null,
                null,
                null,
                null,
                null
        );

        long variantId = variantRepository.insertVariant(command.productId(), variant);

        auditRepository.insertAuditLog(createAuditLog(
                command.userId(),
                "CREATE",
                "variants",
                variantId,
                null,
                String.format("{\"product_id\":%d,\"name\":\"%s\",\"sku\":\"%s\",\"price\":%s,\"cost_price\":%s}",
                        command.productId(), command.name(), command.sku(), command.price(), command.costPrice())
        ));

        if (command.stock() != null && command.stock() > 0) {
            InventoryLot lot = new InventoryLot(null, variantId, null, null, null, command.stock(), command.costPrice(), null, null);
            long lotId = inventoryRepository.insertInventoryLot(lot);

            InventoryAdjustment adjustment = new InventoryAdjustment(null, variantId, lotId, command.stock(), "confirm_receive", null, null, command.userId(), null, null);
            inventoryRepository.insertInventoryAdjustment(adjustment);

            LoggingConfig.getLogger().info("Initial stock {} added for variant {}", command.stock(), variantId);
        }

        return variantId;
    }

    public void updateVariant(UpdateVariantCommand command) {
        if (command.id() == null || command.name() == null || command.price() == null || command.costPrice() == null || command.userId() == null) {
            throw new ValidationException("ID, name, price, cost_price, and userId are required");
        }

        if (command.price().compareTo(BigDecimal.ZERO) < 0 || command.costPrice().compareTo(BigDecimal.ZERO) < 0) {
            throw new ValidationException("Price and cost_price must be non-negative");
        }

        Variant existing = variantRepository.findVariantByIdSync(command.id())
                .orElseThrow(() -> new ValidationException("Variant not found"));

        transactionManager.runInTransaction(() -> {
            updateVariantWithoutTransaction(command, existing);
            return null;
        });
    }

    public void updateVariantWithoutTransaction(UpdateVariantCommand command) {
        Variant existing = variantRepository.findVariantByIdSync(command.id())
                .orElseThrow(() -> new ValidationException("Variant not found"));
        updateVariantWithoutTransaction(command, existing);
    }

    private void updateVariantWithoutTransaction(UpdateVariantCommand command, Variant existing) {
        Variant variant = new Variant(
                command.id(),
                existing.productId(),
                null,
                command.name(),
                command.sku(),
                command.price(),
                command.costPrice(),
                command.stockAlertCap() != null ? command.stockAlertCap() : 10,
                command.isDefault() != null && command.isDefault(),
                command.status() != null ? command.status() : "active",
                null,
                null,
                null,
                null,
                null,
                null,
                null
        );

        variantRepository.updateVariantById(variant);

        auditRepository.insertAuditLog(createAuditLog(
                command.userId(),
                "UPDATE",
                "variants",
                command.id(),
                String.format("{\"name\":\"%s\",\"sku\":\"%s\",\"price\":%s,\"cost_price\":%s}",
                        existing.name(), existing.sku(), existing.price(), existing.costPrice()),
                String.format("{\"name\":\"%s\",\"sku\":\"%s\",\"price\":%s,\"cost_price\":%s}",
                        command.name(), command.sku(), command.price(), command.costPrice())
        ));

        if (command.stock() != null) {
            int targetStock = command.stock();
            if (targetStock >= 0) {
                int currentStock = inventoryRepository.getStockByVariantId(command.id());
                int diff = targetStock - currentStock;

                if (diff != 0) {
                    String reason = command.stockAdjustmentReason() != null ? command.stockAdjustmentReason() : "correction";
                    InventoryAdjustment adjustment = new InventoryAdjustment(null, command.id(), null, diff, reason, null, null, command.userId(), null, null);
                    inventoryRepository.insertInventoryAdjustment(adjustment);
                    LoggingConfig.getLogger().info("Stock adjusted for variant {}: {} (reason: {})",
                            command.id(), (diff > 0 ? "+" : "") + diff, reason);
                }
            }
        }
    }

    public void deleteVariant(long id, long userId) {
        Variant existing = variantRepository.findVariantByIdSync(id)
                .orElseThrow(() -> new ValidationException("Variant not found"));

        int changes = variantRepository.softDeleteVariant(id);

        auditRepository.insertAuditLog(createAuditLog(
                userId,
                "DELETE",
                "variants",
                id,
                String.format("{\"name\":\"%s\",\"sku\":\"%s\"}", existing.name(), existing.sku()),
                null
        ));

        LoggingConfig.getLogger().info("Variant {} soft deleted by user {}", id, userId);
    }

    public PagedResult<Variant> getVariants(VariantFilterCriteria criteria) {
        return variantRepository.findVariants(
                criteria.searchTerm(),
                criteria.categoryId(),
                criteria.categories(),
                criteria.taxCategories(),
                criteria.stockStatus(),
                criteria.status(),
                criteria.sortBy() != null ? criteria.sortBy() : "p.name",
                criteria.sortOrder() != null ? criteria.sortOrder() : "ASC",
                criteria.currentPage() != null ? criteria.currentPage() : 1,
                criteria.itemsPerPage() != null ? criteria.itemsPerPage() : 10
        );
    }

    public Map<String, Object> getVariantStats() {
        return variantRepository.getVariantStats();
    }

    public void validateVariantOwnership(long variantId, long productId) {
        Variant variant = variantRepository.findVariantByIdSync(variantId)
                .orElseThrow(() -> new NotFoundException("Variant " + variantId + " not found"));
        if (!variant.productId().equals(productId)) {
            throw new ValidationException("Variant " + variantId + " does not belong to product " + productId);
        }
    }

    private com.possum.domain.model.AuditLog createAuditLog(long userId, String action, String tableName, long rowId, String oldData, String newData) { return new com.possum.domain.model.AuditLog(null, userId, action, tableName, rowId, oldData, newData, null, null, null);
    }

    public record AddVariantCommand(
            Long productId,
            String name,
            String sku,
            BigDecimal price,
            BigDecimal costPrice,
            Integer stockAlertCap,
            Boolean isDefault,
            String status,
            Integer stock,
            Long userId
    ) {}

    public record UpdateVariantCommand(
            Long id,
            String name,
            String sku,
            BigDecimal price,
            BigDecimal costPrice,
            Integer stockAlertCap,
            Boolean isDefault,
            String status,
            Integer stock,
            String stockAdjustmentReason,
            Long userId
    ) {}

    public record VariantFilterCriteria(
            String searchTerm,
            Long categoryId,
            List<Long> categories,
            List<Long> taxCategories,
            List<String> stockStatus,
            List<String> status,
            String sortBy,
            String sortOrder,
            Integer currentPage,
            Integer itemsPerPage
    ) {}
}
